libraries/lib-music-information-retrieval/StftFrameProvider.cpp

```diff
@@ -0,0 +1,91 @@
+#include "StftFrameProvider.h"
+#include "MirAudioReader.h"
+#include "MirUtils.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <numeric>
+
+namespace MIR
+{
+namespace
+{
+constexpr auto twoPi = 2 * 3.14159265358979323846;
+
+int GetFrameSize(int sampleRate)
+{
+   // 2048 frame size for sample rate 44.1kHz
+   return 1 << (11 + (int)std::round(std::log2(sampleRate / 44100.)));
+}
+
+double GetHopSize(int sampleRate, long long numSamples)
+{
+   // Aim for a hop size closest to 10ms, yet dividing `numSamples` to a power
+   // of two. This will spare us the need for resampling when we need to get the
+   // autocorrelation of the ODF using an FFT.
+   const auto idealHopSize = 0.01 * sampleRate;
+   const auto numFrames =
+      1 << (int)std::round(std::log2(numSamples / idealHopSize));
+   return 1. * numSamples / numFrames;
+}
+} // namespace
+
+StftFrameProvider::StftFrameProvider(const MirAudioReader& audio)
+    : mAudio { audio }
+    , mFftSize { GetFrameSize(audio.GetSampleRate()) }
+    , mHopSize { GetHopSize(audio.GetSampleRate(), audio.GetNumSamples()) }
+    , mWindow { GetNormalizedHann(mFftSize) }
+    , mNumFrames { static_cast<int>(
+         std::round(audio.GetNumSamples() / mHopSize)) }
+    , mNumSamples { audio.GetNumSamples() }
+{
+   assert(IsPowOfTwo(mNumFrames));
+}
+
+bool StftFrameProvider::GetNextFrame(std::vector<float>& frame)
+{
+   if (mNumFramesProvided >= mNumFrames)
+      return false;
+   frame.resize(mFftSize);
```

Comment:
I figured out that the only uses of `GetFloats` are in this function.  Is there a guarantee that `frame.data()` used as the buffer will be zero-initialized first?  Only if the given `frame` is empty before this `resize`.  Precede this with `resize(0)` and then there is no more doubt.

Reply:
https://github.com/audacity/audacity/pull/5791#discussion_r1437562009